### ISSUE TITLE
conv1d/conv2d silently ignore conv_config.activation on the 1D-depthwise path: is_1d_depthwise_conv excludes has_bias but not activation, and the depthwise kernel cannot apply one

### ISSUE THREAD

**Issue by singhharsh1708** [tenstorrent/tt-metal#55272]
### Summary

`ttnn.conv1d` (and `ttnn.conv2d` on the 1D-depthwise path) silently ignores `conv_config.activation`. The 1D-depthwise kernel cannot fuse an activation, and the guard that routes work to it excludes `has_bias` but not activation, so a depthwise conv with an activation and no bias lands on a kernel that never applies it. No error is raised and the activation is not applied anywhere else.

### The guard has a term missing

`ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:536-546`:

```cpp
bool is_1d_depthwise_conv(
    uint32_t groups,
    uint32_t input_channels,
    uint32_t output_channels,
    uint32_t kernel_height,
    uint32_t image_height,
    bool has_bias) {
    bool is_depthwise_conv = groups == input_channels && groups == output_channels;
    ...
    return is_depthwise_conv && is_1d_conv(kernel_height, image_height) && !has_bias;
}
```

`!has_bias` is there because the specialized kernel cannot fuse a bias. It equally cannot fuse an activation, and there is no corresponding term.

### The selected kernel has no activation support

When the predicate is true, the factory selects `compute_depthwise_conv1d.cpp` (`conv2d_op_sharded_program_factory.cpp:793-795`). That kernel contains no activation handling at all:

```
$ grep -c "SFPU_OP_INIT_ACTIVATION\|SFPU_OP_FUNC_ACTIVATION\|pack_relu" \
    ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/compute_depthwise_conv1d.cpp
0
```

Only the generic path implements them:

```
$ grep -rln "SFPU_OP_FUNC_ACTIVATION" ttnn/cpp/ttnn/operations/conv/
ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize.cpp
```

The host still computes `pack_relu` and builds the SFPU activation defines (`conv2d_op_sharded_program_factory.cpp:919-923`) and attaches them (`:1203-1205`), but the depthwise compute kernel's compile-time args stop at index 12 with no `pack_relu` (`:1044-1096`), where the non-depthwise branch passes it at `:1128`. The defines are dead for this kernel.

### Nothing re-applies it afterwards

```
$ grep -rn "unary_chain\|apply_activation" ttnn/cpp/ttnn/operations/conv/
(no matches)
```

This is where it differs from `ttnn.matmul`, which falls back to `ttnn::unary_chain` when it cannot fuse (`matmul.cpp:355-359`). Conv has no such fallback, so the activation is simply lost.

### conv1d always satisfies the shape condition

`conv1d.cpp:38-55` reshapes the input to `[N, 1, L, C]` and reinterprets the weight as `[..., 1, kernel_size]`, so `image_height == 1` and `kernel_height == 1` always. Any depthwise `conv1d` (`groups == in_channels == out_channels`) without a bias therefore takes this path.

### Reproducing

```python
out = ttnn.conv1d(
    input_tensor=x,                     # [1, 64, 32] bf16
    weight_tensor=w,                    # [32, 1, 3]
    bias_tensor=None,
    in_channels=32, out_channels=32, groups=32,
    kernel_size=3, stride=1, padding=1, dilation=1,
    conv_config=ttnn.Conv1dConfig(activation=ttnn.UnaryWithParam(ttnn.UnaryOpType.RELU)),
    ...
)
```

Expected `relu(conv1d(x, w))` with every negative pre-activation clamped to 0; actual is the raw convolution, negatives included. Adding a bias, or using `kernel_height > 1`, makes the predicate false and the activation is applied correctly — which is the clearest way to see the difference.

`Conv2dConfig::activation` is a public property (`conv2d_nanobind.cpp:294-297`, documented "Fused activation function to be applied on the output"), and `Conv1dConfig` is an alias of `Conv2dConfig` (`conv1d.hpp:17`).

### Magnitude

Roughly half the outputs are wrong by their full magnitude rather than by a rounding step. Simulating the shape above (C=32, L=64, K=3, `x ~ N(0,1)`, `w ~ N(0,1/sqrt(3))`, 2048 outputs) against a torch golden gives PCC 0.843 for RELU, 0.882 for SILU, 0.841 for GELU. Conv tests in this repo assert PCC >= 0.99, so this is far outside the tolerance tier rather than a metric technicality.

### Why no test catches it

No in-tree test crosses both conditions:

- `tests/ttnn/nightly/unit_tests/operations/conv/test_conv2d.py::test_conv_activation` (704-793) sweeps RELU/SILU/SIGMOID/TANH/SQRT/GELU but pins `filter=3` and `has_bias=True`, so the predicate is false on both counts.
- `test_conv_packer_l1_acc_untilize_activation_without_bias` (:655) is bias-free with GELU but uses a 3x3 kernel on a 32x32 image.
- The two depthwise conv1d tests (`test_conv1d.py:23`, `:80`) never set an activation — the only `activation` token in that file is `deallocate_activation=False`.

### Fix

Add the missing term to the existing guard, mirroring `has_bias`. At `conv2d_utils.cpp:545`, exclude a fused activation as well — rename the last parameter to `has_bias_or_activation` and pass `has_bias || conv_config.activation.has_value()` at the call sites (`conv2d.cpp:160-161`, `conv2d_utils.cpp:1041-1042` and `:1215-1216`, `prepare_conv2d_weights.cpp:1226-1227` and `:1514-1515`), plus `operation_attributes.has_bias || operation_attributes.activation.has_value()` at `conv2d_op_sharded_program_factory.cpp:334-335`, where the predicate is recomputed.

That routes these convs onto `conv_bmm_tilize.cpp`, which already applies the activation and already handles depthwise-shaped grouped convs — exactly what happens today as soon as a bias is present.

Alternatively, implement the activation in `compute_depthwise_conv1d.cpp`; the guard fix is the smaller change and matches how bias is already handled.

Measured on **ttsim** (tenstorrent/ttsim v1.10.3, `libttsim_bh.so`, Blackhole) with the prebuilt tt-metal `v0.79.0-dev20260902` ubuntu22.04 debs and SFPI 7.72.0. No hardware is required to reproduce this. Every file:line is against `2aa657a`.

Depthwise `conv1d` (groups == in_channels == out_channels = 32, kernel 3, no bias), run twice with only `conv_config.activation` changing:

```
act=None: negatives=1018/2048  min=-2.45312
act=RELU: negatives=1018/2048  min=-2.45312
```

The two outputs are identical — 1018 of 2048 elements remain negative with RELU requested, and the minimum is unchanged. The activation is not applied on this path.

### The same call applies the activation correctly one flag away

Identical inputs, identical `conv_config`, identical everything except `groups`:

| configuration | `activation=None` | `activation=RELU` |
|---|---|---|
| depthwise (`groups = in_channels = 32`) | 1073 negatives / 2048 | **1073 negatives / 2048** |
| dense (`groups = 1`) | 1053 negatives / 2048 | **0 negatives / 2048** |

The dense path applies RELU exactly as expected — every negative is clamped away. The depthwise path returns a bit-identical result with and without it.

So this is not a limitation of conv, of the activation plumbing, or of the config: the op honours `conv_config.activation` correctly whenever `is_1d_depthwise_conv()` is false, and drops it silently whenever that predicate is true. Setting `groups = 1` on the same call is enough to make the activation reappear.

I would like to work on this through the bounty program, and am happy to prepare the PR.

**Comment by singhharsh1708** [comment]
Measured on **ttsim** (tenstorrent/ttsim v1.10.3, `libttsim_bh.so`, Blackhole) with the prebuilt tt-metal `v0.79.0-dev20260902` ubuntu22.04 debs and SFPI 7.72.0 — no hardware.

Depthwise `conv1d` (groups == in_channels == out_channels = 32, kernel 3, no bias), run twice with only `conv_config.activation` changing:

```
act=None: negatives=1018/2048  min=-2.45312
act=RELU: negatives=1018/2048  min=-2.45312
```

The two outputs are identical — 1018 of 2048 elements remain negative with RELU requested, and the minimum is unchanged. The activation is not applied on this path.

**Comment by borjamoskv** [comment]
/assign

### Analytical Root Cause & Deterministic Fix

**Root Cause:**
In `ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:535-546`, `is_1d_depthwise_conv` checks `!has_bias` because the specialized 1D depthwise kernel (`compute_depthwise_conv1d.cpp`) cannot fuse a bias. However, it completely omits checking whether `conv_config.activation` is configured. As a result, 1D depthwise convolutions with activations (and no bias) are silently routed to `compute_depthwise_conv1d.cpp`, which lacks activation support, silently dropping the requested activation with zero error.

**Deterministic Resolution:**
1. In `conv2d_utils.hpp` and `conv2d_utils.cpp`, extend `is_1d_depthwise_conv` with `bool has_activation = false`:
```cpp
bool is_1d_depthwise_conv(
    uint32_t groups,
    uint32_t input_channels,
    uint32_t output_channels,
    uint32_t kernel_height,
    uint32_t image_height,
    bool has_bias,
    bool has_activation = false) {
    bool is_depthwise_conv = groups == input_channels && groups == output_channels;
    return is_depthwise_conv && is_1d_conv(kernel_height, image_height) && !has_bias && !has_activation;
}
```
2. In `conv2d.cpp:161`, pass `conv_config.activation.has_value()` to ensure convolutions with activations route to the general conv kernel that supports fused activations.

**Validation & Branch:**
Pre-staged branch ready in fork: `borjamoskv:fix/conv2d-depthwise-activation-guard`.
Awaiting assignment to dispatch PR immediately.
